### ISSUE TITLE
Unable to expose types from submodules in the public API of jsii modules

### ISSUE THREAD

**Issue by eladb** [aws/jsii#1629]
### Affected Languages

- [x] `TypeScript` or `Javascript`
- [x] `Python`
- [x] `Java`
- [x] .NET (`C#`, `F#`, ...)

### General Information
* **JSII Version:** 1.4.1

### What is the problem?

Consider a jsii library which depends on another jsii library that uses submodules (e.g. `monocdk-experiment`) and the consuming library exposes a type from a submodule in it's own public API:

```ts
import { Construct } from 'monocdk-experiment';
import { aws_s3 } from 'monocdk-experiment';
import { Topic } from 'monocdk-experiment/aws-sns';

export class MyConstruct extends Construct {
  public readonly myBucket: aws_s3.Bucket;
  public readonly myTopic: Topic;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    this.myBucket = new aws_s3.Bucket(this, 'MyBucket');
    this.myTopic = new Topic(this, 'MyTopic');
  }
}
```

A full repro is available at https://github.com/eladb/jsii-submodule-consumption-repro:

```
git clone https://github.com/eladb/jsii-submodule-consumption-repro.git
cd jsii-submodule-consumption-repro
yarn install
yarn build
```

### Verbose Log

The jsii compiler fails with the following output:

```
[2020-05-05T10:47:00.799] [ERROR] jsii/compiler - Type model errors prevented the JSII assembly from being created
node_modules/monocdk-experiment/aws-s3/lib/bucket.d.ts:695:1 - error TS9999: JSII: Unable to resolve referenced type 'monocdk-experiment.Bucket'. Type may be @internal or unexported

695 export declare class Bucket extends BucketBase {
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
696     static fromBucketArn(scope: Construct, id: string, bucketArn: string): IBucket;
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
...
803     private allowLogDelivery;
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
804 }
    ~
node_modules/monocdk-experiment/aws-s3/lib/bucket.d.ts:695:1 - error TS9999: JSII: Use of foreign type not present in the monocdk-experiment's assembly: monocdk-experiment.Bucket

695 export declare class Bucket extends BucketBase {
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
696     static fromBucketArn(scope: Construct, id: string, bucketArn: string): IBucket;
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
...
803     private allowLogDelivery;
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
804 }
    ~
node_modules/monocdk-experiment/aws-sns/lib/topic.d.ts:34:1 - error TS9999: JSII: Unable to resolve referenced type 'monocdk-experiment.Topic'. Type may be @internal or unexported

 34 export declare class Topic extends TopicBase {
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 35     static fromTopicArn(scope: Construct, id: string, topicArn: string): ITopic;
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
...
 39     constructor(scope: Construct, id: string, props?: TopicProps);
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 40 }
    ~
node_modules/monocdk-experiment/aws-sns/lib/topic.d.ts:34:1 - error TS9999: JSII: Use of foreign type not present in the monocdk-experiment's assembly: monocdk-experiment.Topic

 34 export declare class Topic extends TopicBase {
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 35     static fromTopicArn(scope: Construct, id: string, topicArn: string): ITopic;
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
...
 39     constructor(scope: Construct, id: string, props?: TopicProps);
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 40 }
    ~
```

It appears that the compiler does not resolve the FQNs of these types to their full submodule location.

Note that this repro includes both "barrel imports" (`import { Topic } from 'monocdk-experiment/aws-sns`) and namespaced imported (`import { aws_s3 } from 'monocdk-experiment'`) and both types produce the same error.

**Comment by eladb** [comment]
Copy @NetaNir

**Comment by MrArnoldPalmer** [comment]
I'm wondering why `jsii-calc` isn't showing this error for [this test case](https://github.com/aws/jsii/blob/master/packages/jsii-calc/lib/submodules.ts#L6). Building the same class using barrel imports like so DOES show the issue though.

```typescript
import { IReflectable, ReflectableEntry, Reflector } from '@scope/jsii-calc-lib/lib/submodule';

export class UpcasingReflectableNamed implements IReflectable {
  public static readonly reflector = new Reflector();

  public constructor(private readonly delegate: Record<string, unknown>) { }

  public get entries(): ReflectableEntry[] {
      return Object.entries(this.delegate)
          .map(([key, value]) => ({ key: key.toLocaleUpperCase(), value }));
  }
}
```
 Developing against the repro-repo for now to see if we can shed some light on that.

**Comment by MrArnoldPalmer** [comment]
Fixed by #1557 and released in v1.5.0. Current version of monocdk-experiment built with 1.4.1 which is why the repro-repo throws the error.
